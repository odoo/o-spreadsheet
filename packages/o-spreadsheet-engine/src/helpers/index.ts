export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
