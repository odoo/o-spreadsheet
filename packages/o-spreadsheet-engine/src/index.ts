export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
