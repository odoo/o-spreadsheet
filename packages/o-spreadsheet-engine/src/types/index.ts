export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
