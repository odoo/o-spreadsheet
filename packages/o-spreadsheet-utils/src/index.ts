export * from "./translation";
export * from "./uuid";
