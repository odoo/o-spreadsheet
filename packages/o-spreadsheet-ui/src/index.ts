import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}