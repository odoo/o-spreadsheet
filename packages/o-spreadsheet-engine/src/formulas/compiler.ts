import { parseNumber, unquote } from "../helpers";
import { _t } from "../translation";
import { CompiledFormula, FormulaToExecute } from "../types/base";
import { BadExpressionError, UnknownFunctionError } from "../types/errors";
import { DEFAULT_LOCALE } from "../types/locale";
import { FunctionCode, FunctionCodeBuilder, Scope } from "./code_builder";
import { parseTokens } from "./parser";
import { rangeTokenize } from "./range_tokenizer";
import { Token } from "./tokenizer";

interface ArgDefinitionLike {
  type?: string[];
}

interface FunctionDescriptionLike {
  args: ArgDefinitionLike[];
  minArgRequired: number;
  maxArgPossible: number;
  nbrArgRepeating: number;
  nbrArgOptional: number;
}

interface FunctionRegistryLike {
  content: Record<string, FunctionDescriptionLike>;
}

type ArgTargetingFn = (
  fn: FunctionDescriptionLike,
  argCount: number
) => (argIndex: number) => number | undefined;

let functionRegistryProvider: (() => FunctionRegistryLike) | null = null;
let argTargetingImpl: ArgTargetingFn | null = null;

export function setFunctionRegistryProvider(provider: () => FunctionRegistryLike) {
  functionRegistryProvider = provider;
}

export function setArgTargetingImplementation(fn: ArgTargetingFn) {
  argTargetingImpl = fn;
}

function getFunctionRegistry(): FunctionRegistryLike {
  if (!functionRegistryProvider) {
    throw new Error("Function registry provider has not been set.");
  }
  return functionRegistryProvider();
}

function getArgTargeting(): ArgTargetingFn {
  if (!argTargetingImpl) {
    throw new Error("argTargeting implementation has not been set.");
  }
  return argTargetingImpl;
}

export const OPERATOR_MAP = {
  "=": "EQ",
  "+": "ADD",
  "-": "MINUS",
  "*": "MULTIPLY",
  "/": "DIVIDE",
  ">=": "GTE",
  "<>": "NE",
  ">": "GT",
  "<=": "LTE",
  "<": "LT",
  "^": "POWER",
  "&": "CONCATENATE",
} as const;

export const UNARY_OPERATOR_MAP = {
  "-": "UMINUS",
  "+": "UPLUS",
  "%": "UNARY.PERCENT",
} as const;

interface LiteralValues {
  numbers: { value: number }[];
  strings: { value: string }[];
}

type InternalCompiledFormula = CompiledFormula & {
  literalValues: LiteralValues;
  symbols: string[];
};

export const functionCache: Record<string, FormulaToExecute> = {};

export function compile(formula: string): CompiledFormula {
  const tokens = rangeTokenize(formula);
  return compileTokens(tokens);
}

export function compileTokens(tokens: Token[]): CompiledFormula {
  try {
    return compileTokensOrThrow(tokens);
  } catch (error) {
    return {
      tokens,
      dependencies: [],
      execute() {
        return error;
      },
      isBadExpression: true,
      normalizedFormula: tokens.map((t) => t.value).join(""),
    };
  }
}

function compileTokensOrThrow(tokens: Token[]): CompiledFormula {
  const registry = getFunctionRegistry();
  const functions = registry.content;
  const argTargeting = getArgTargeting();

  const { dependencies, literalValues, symbols } = formulaArguments(tokens);
  const cacheKey = compilationCacheKey(tokens);
  if (!functionCache[cacheKey]) {
    const ast = parseTokens([...tokens]);
    const scope = new Scope();

    let stringCount = 0;
    let numberCount = 0;
    let dependencyCount = 0;

    if (ast.type === "BIN_OPERATION" && ast.value === ":") {
      throw new BadExpressionError(_t("Invalid formula"));
    }
    if (ast.type === "EMPTY") {
      throw new BadExpressionError(_t("Invalid formula"));
    }

    const compiledAST = compileAST(ast);
    const code = new FunctionCodeBuilder();
    code.append(`// ${cacheKey}`);
    code.append(compiledAST);
    code.append(`return ${compiledAST.returnExpression};`);
    const baseFunction = new Function(
      "deps",
      "ref",
      "range",
      "getSymbolValue",
      "ctx",
      code.toString()
    );

    // @ts-ignore - constructed function
    functionCache[cacheKey] = baseFunction;

    function compileFunctionArgs(astFuncall: any): FunctionCode[] {
      const { args } = astFuncall;
      const functionName = astFuncall.value.toUpperCase();
      const functionDefinition = functions[functionName];

      if (!functionDefinition) {
        throw new UnknownFunctionError(_t('Unknown function: "%s"', astFuncall.value));
      }

      assertEnoughArgs(args, functionDefinition, functionName);

      const compiledArgs: FunctionCode[] = [];
      const argToFocus = argTargeting(functionDefinition, args.length);

      for (let i = 0; i < args.length; i++) {
        const argIndex = argToFocus(i) ?? -1;
        const argDefinition =
          functionDefinition.args[argIndex] ??
          functionDefinition.args[i] ??
          ({} as ArgDefinitionLike);
        const currentArg = args[i];
        const argTypes = argDefinition?.type || [];
        const isMeta = argTypes.includes("META") || argTypes.includes("RANGE<META>");
        const hasRange = argTypes.some((t: string) => isRangeType(t));
        compiledArgs.push(compileAST(currentArg, isMeta, hasRange));
      }

      return compiledArgs;
    }

    function compileAST(ast: any, isMeta = false, hasRange = false): FunctionCode {
      const codeBuilder = new FunctionCodeBuilder(scope);
      if (ast.type !== "REFERENCE" && !(ast.type === "BIN_OPERATION" && ast.value === ":")) {
        if (isMeta) {
          throw new BadExpressionError(_t("Argument must be a reference to a cell or range."));
        }
      }
      if (ast.debug) {
        codeBuilder.append("debugger;");
        codeBuilder.append(`ctx["debug"] = true;`);
      }
      switch (ast.type) {
        case "BOOLEAN":
          return codeBuilder.return(`{ value: ${ast.value} }`);
        case "NUMBER":
          return codeBuilder.return(`this.literalValues.numbers[${numberCount++}]`);
        case "STRING":
          return codeBuilder.return(`this.literalValues.strings[${stringCount++}]`);
        case "REFERENCE":
          return codeBuilder.return(
            `${ast.value.includes(":") || hasRange ? "range" : "ref"}(deps[${dependencyCount++}], ${
              isMeta ? "true" : "false"
            })`
          );
        case "FUNCALL": {
          const args = compileFunctionArgs(ast).map((argCode) => argCode.assignResultToVariable());
          codeBuilder.append(...args);
          const fnName = ast.value.toUpperCase();
          return codeBuilder.return(`ctx['${fnName}'](${args.map((arg) => arg.returnExpression)})`);
        }
        case "UNARY_OPERATION": {
          const fnName = UNARY_OPERATOR_MAP[ast.value as keyof typeof UNARY_OPERATOR_MAP];
          const operand = compileAST(ast.operand, false, false).assignResultToVariable();
          codeBuilder.append(operand);
          return codeBuilder.return(`ctx['${fnName}'](${operand.returnExpression})`);
        }
        case "BIN_OPERATION": {
          const fnName = OPERATOR_MAP[ast.value as keyof typeof OPERATOR_MAP];
          const left = compileAST(ast.left, false, false).assignResultToVariable();
          const right = compileAST(ast.right, false, false).assignResultToVariable();
          codeBuilder.append(left);
          codeBuilder.append(right);
          return codeBuilder.return(
            `ctx['${fnName}'](${left.returnExpression}, ${right.returnExpression})`
          );
        }
        case "SYMBOL": {
          const symbolIndex = symbols.indexOf(ast.value);
          return codeBuilder.return(`getSymbolValue(this.symbols[${symbolIndex}])`);
        }
        case "EMPTY":
          return codeBuilder.return("undefined");
        default:
          return codeBuilder.return("undefined");
      }
    }
  }

  const compiledFormula: InternalCompiledFormula = {
    execute: functionCache[cacheKey],
    dependencies,
    literalValues,
    symbols,
    tokens,
    isBadExpression: false,
    normalizedFormula: cacheKey,
  };
  return compiledFormula;
}

function compilationCacheKey(tokens: Token[]): string {
  let cacheKey = "";
  for (const token of tokens) {
    switch (token.type) {
      case "STRING":
        cacheKey += "|S|";
        break;
      case "NUMBER":
        cacheKey += "|N|";
        break;
      case "REFERENCE":
      case "INVALID_REFERENCE":
        cacheKey += token.value.includes(":") ? "|R|" : "|C|";
        break;
      case "SPACE":
        break;
      default:
        cacheKey += token.value;
    }
  }
  return cacheKey;
}

function formulaArguments(tokens: Token[]) {
  const literalValues: LiteralValues = {
    numbers: [],
    strings: [],
  };
  const dependencies: string[] = [];
  const symbols: string[] = [];
  for (const token of tokens) {
    switch (token.type) {
      case "INVALID_REFERENCE":
      case "REFERENCE":
        dependencies.push(token.value);
        break;
      case "STRING":
        literalValues.strings.push({ value: unquote(token.value) });
        break;
      case "NUMBER":
        literalValues.numbers.push({ value: parseNumber(token.value, DEFAULT_LOCALE) });
        break;
      case "SYMBOL":
        symbols.push(unquote(token.value, "'"));
        break;
      default:
        break;
    }
  }
  return { dependencies, literalValues, symbols };
}

function assertEnoughArgs(
  args: any[],
  functionDefinition: FunctionDescriptionLike,
  functionName: string
) {
  const nbrArgSupplied = args.length;
  const { nbrArgRepeating, minArgRequired } = functionDefinition;

  if (nbrArgSupplied < minArgRequired) {
    throw new BadExpressionError(
      _t(
        "Invalid number of arguments for the %s function. Expected %s minimum, but got %s instead.",
        functionName,
        minArgRequired,
        nbrArgSupplied
      )
    );
  }

  if (nbrArgSupplied > functionDefinition.maxArgPossible) {
    throw new BadExpressionError(
      _t(
        "Invalid number of arguments for the %s function. Expected %s maximum, but got %s instead.",
        functionName,
        functionDefinition.maxArgPossible,
        nbrArgSupplied
      )
    );
  }

  if (nbrArgRepeating > 1) {
    const nbrValueRepeating =
      nbrArgRepeating * Math.floor((nbrArgSupplied - minArgRequired) / nbrArgRepeating);
    const nbrValueRemaining =
      nbrArgSupplied - minArgRequired - nbrValueRepeating - functionDefinition.nbrArgOptional;

    if (nbrValueRemaining > 0) {
      throw new BadExpressionError(
        _t(
          "Invalid number of arguments for the %s function. Repeatable arguments should be supplied in groups of %s, with up to %s optional. Got %s too many.",
          functionName,
          nbrArgRepeating,
          functionDefinition.nbrArgOptional,
          nbrValueRemaining
        )
      );
    }
  }
}

function isRangeType(type: string) {
  return type.startsWith("RANGE");
}
