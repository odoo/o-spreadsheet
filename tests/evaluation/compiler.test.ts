import { Model } from "../../src";
import { functionCache } from "../../src/formulas/compiler";
import { compile } from "../../src/formulas/index";
import { functionRegistry } from "../../src/functions";
import { createValidRange } from "../../src/helpers";
import { CompiledFormula } from "../../src/types";
import { evaluateCell, evaluateCellFormat, restoreDefaultFunctions } from "../test_helpers/helpers";

function compiledBaseFunction(formula: string): CompiledFormula {
  for (let f in functionCache) {
    delete functionCache[f];
  }
  return compileFromCompleteFormula(formula);
}

function compileFromCompleteFormula(formula: string) {
  return compile(formula);
}

describe("expression compiler", () => {
  test.each(["=1", "=true", `="abc"`])("some arithmetic expressions", (formula) => {
    const compiledFormula = compiledBaseFunction(formula);
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("simple values that throw error", () => {
    expect(() => compiledBaseFunction(`='abc'`)).toThrowError();
  });

  test.each(["=1 + 3", "=2 * 3", "=2 - 3", "=2 / 3", "=-3", "=(3 + 1) * (-1 + 4)"])(
    "some arithmetic expressions",
    (formula) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.execute.toString()).toMatchSnapshot();
    }
  );

  test.each(["=sum(1,2)", '=sum(true, "")', "=sum(1,,2)"])(
    "some arithmetic expressions",
    (formula) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.execute.toString()).toMatchSnapshot();
    }
  );

  test.each(["=1%", "=(2+5)%", "=A1%"])("some arithmetic expressions", (formula) => {
    const compiledFormula = compiledBaseFunction(formula);
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("read some values and functions", () => {
    const compiledFormula = compiledBaseFunction("=A1 + sum(A2:C3)");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("with the same reference multiple times", () => {
    const compiledFormula = compiledBaseFunction("=SUM(A1, A1, A2)");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("expression with $ref", () => {
    const compiledFormula = compiledBaseFunction("=$A1+$A$2+A$3");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("expression with references with a sheet", () => {
    const compiledFormula = compiledBaseFunction("=Sheet34!B3");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("expressions with a debugger", () => {
    const compiledFormula = compiledBaseFunction("=? A1 / 2");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("cells are converted to ranges if function require a range", () => {
    const compiledFormula = compiledBaseFunction("=sum(A1)");
    expect(compiledFormula.execute.toString()).toMatchSnapshot();
  });

  test("cannot compile some invalid formulas", () => {
    expect(() => compiledBaseFunction("=qsdf")).toThrow();
  });
});

describe("compile functions", () => {
  describe("check number of arguments", () => {
    afterAll(() => {
      restoreDefaultFunctions();
    });

    test("with basic arguments", () => {
      functionRegistry.add("ANYFUNCTION", {
        description: "any function",
        compute: () => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"] },
        ],
      });
      expect(() => compiledBaseFunction("=ANYFUNCTION()")).toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1)")).toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1,2,3)")).toThrow();
      restoreDefaultFunctions();
    });

    test("with optional argument", () => {
      functionRegistry.add("OPTIONAL", {
        description: "function with optional argument",
        compute: () => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true },
        ],
      });
      expect(() => compiledBaseFunction("=OPTIONAL(1)")).not.toThrow();
      expect(() => compiledBaseFunction("=OPTIONAL(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=OPTIONAL(1,2,3)")).toThrow();
      restoreDefaultFunctions();
    });

    test("with default argument", () => {
      functionRegistry.add("USEDEFAULTARG", {
        description: "function with a default argument",
        compute: () => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], default: true, defaultValue: 42 },
        ],
      });
      expect(() => compiledBaseFunction("=USEDEFAULTARG(1)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEDEFAULTARG(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEDEFAULTARG(1,2,3)")).toThrow();
      restoreDefaultFunctions();
    });

    test("with repeatable argument", () => {
      functionRegistry.add("REPEATABLE", {
        description: "function with repeatable argument",
        compute: (arg) => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true, repeating: true },
        ],
      });
      expect(() => compiledBaseFunction("=REPEATABLE(1)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLE(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLE(1,2,3,4,5,6)")).not.toThrow();
      restoreDefaultFunctions();
    });

    test("with more than one repeatable argument", () => {
      functionRegistry.add("REPEATABLES", {
        description: "any function",
        compute: (arg) => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true, repeating: true },
          { name: "arg3", description: "", type: ["ANY"], optional: true, repeating: true },
        ],
      });
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2)")).toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3, 4)")).toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3, 4, 5)")).not.toThrow();
      restoreDefaultFunctions();
    });
  });

  describe("interpret arguments", () => {
    beforeAll(() => {
      functionRegistry.add("ISSECONDARGUNDEFINED", {
        description: "any function",
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"] },
        ],
        compute: (arg1, arg2) => {
          return {
            value: arg2 === undefined,
            format: arg2 === undefined ? "TRUE" : "FALSE",
          };
        },
      });

      functionRegistry.add("SECONDARGDEFAULTVALUEEQUAL42", {
        description: "function with a default argument",
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], default: true, defaultValue: 42 },
        ],
        compute: (arg1, arg2 = { value: 42, format: "42" }) => {
          return !Array.isArray(arg2) && arg2.value === 42 && arg2.format === "42"
            ? { value: true, format: "TRUE" }
            : { value: false, format: "FALSE" };
        },
      });
    });
    afterAll(() => {
      restoreDefaultFunctions();
    });
    test("empty value interpreted as undefined", () => {
      expect(evaluateCell("A1", { A1: "=ISSECONDARGUNDEFINED(1,)" })).toBe(true);
      expect(evaluateCellFormat("A1", { A1: "=ISSECONDARGUNDEFINED(1,)" })).toBe("TRUE");
    });

    test("if default value --> empty value interpreted as default value", () => {
      expect(evaluateCell("A1", { A1: "=SECONDARGDEFAULTVALUEEQUAL42(1,)" })).toBe(true);
      expect(evaluateCellFormat("A1", { A1: "=SECONDARGDEFAULTVALUEEQUAL42(1,)" })).toBe("TRUE");
    });

    test("if default value --> non-value interpreted as default value", () => {
      expect(evaluateCell("A1", { A1: "=SECONDARGDEFAULTVALUEEQUAL42(1)" })).toBe(true);
      expect(evaluateCellFormat("A1", { A1: "=SECONDARGDEFAULTVALUEEQUAL42(1)" })).toBe("TRUE");
    });
  });

  describe("with meta arguments", () => {
    beforeAll(() => {
      functionRegistry.add("USEMETAARG", {
        description: "function with a meta argument",
        compute: () => true,
        args: [{ name: "arg", description: "", type: ["META"] }],
      });
      functionRegistry.add("NOTUSEMETAARG", {
        description: "any function",
        compute: () => true,
        args: [{ name: "arg", description: "", type: ["ANY"] }],
      });
    });

    test.each(["=USEMETAARG(A1)", "=USEMETAARG(B2)"])(
      "function call requesting meta parameter",
      (formula) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.execute.toString()).toMatchSnapshot();
      }
    );

    test("throw error if parameter isn't cell/range reference", () => {
      expect(() => compiledBaseFunction("=USEMETAARG(X8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG($X$8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG(Sheet42!X8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG('Sheet 42'!X8)")).not.toThrow();

      expect(() => compiledBaseFunction("=USEMETAARG(D3:Z9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG($D$3:$Z$9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG(Sheet42!$D$3:$Z$9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG('Sheet 42'!D3:Z9)")).not.toThrow();

      expect(() => compiledBaseFunction('=USEMETAARG("kikou")')).toThrowError();
      expect(() => compiledBaseFunction('=USEMETAARG("")')).toThrowError();
      expect(() => compiledBaseFunction("=USEMETAARG(TRUE)")).toThrowError();
      expect(() => compiledBaseFunction("=USEMETAARG(SUM(1,2,3))")).toThrowError();
    });

    test("do not care about the value of the cell / range passed as a reference", () => {
      const compiledFormula1 = compileFromCompleteFormula("=USEMETAARG(A1)");
      const compiledFormula2 = compileFromCompleteFormula("=USEMETAARG(A1:B2)");
      const compiledFormula3 = compileFromCompleteFormula("=NOTUSEMETAARG(A1)");
      const compiledFormula4 = compileFromCompleteFormula("=NOTUSEMETAARG(A1:B2)");

      const m = new Model();

      let refFn = jest.fn();
      let ensureRange = jest.fn();

      const ctx = { USEMETAARG: () => {}, NOTUSEMETAARG: () => {} };

      const rangeA1 = createValidRange(m.getters, "ABC", "A1")!;
      const rangeA1ToB2 = createValidRange(m.getters, "ABC", "A1:B2")!;

      compiledFormula1.execute([rangeA1], refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(rangeA1, true);
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula2.execute([rangeA1ToB2], refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(rangeA1ToB2, true);
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula3.execute([rangeA1], refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(rangeA1, false);
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula4.execute([rangeA1ToB2], refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledTimes(0);
      expect(ensureRange).toHaveBeenCalledWith(rangeA1ToB2);
      refFn.mockReset();
    });
  });

  test("function cache ignore spaces in functions", () => {
    compiledBaseFunction("=SUM(A1)");
    expect(Object.keys(functionCache)).toEqual(["=SUM(C|0|)"]);
    compile("= SUM(A1)");
    compile("=SUM( A1)");
    compile("= SUM(A1 )");
    compile("= SUM   (    A1    )");
    expect(Object.keys(functionCache)).toEqual(["=SUM(C|0|)"]);
  });
});
