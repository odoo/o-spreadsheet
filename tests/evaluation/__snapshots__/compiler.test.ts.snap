// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile functions with meta arguments function call requesting meta parameter 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =USEMETAARG(|0|)
const _1 = ref(deps[0], true, "USEMETAARG",  1);
return ctx['USEMETAARG'](_1);
}"
`;

exports[`compile functions with meta arguments function call requesting meta parameter 2`] = `
"function anonymous(deps,ref,range,ctx
) {
// =USEMETAARG(|0|)
const _1 = ref(deps[0], true, "USEMETAARG",  1);
return ctx['USEMETAARG'](_1);
}"
`;

exports[`expression compiler cells are converted to ranges if function require a range 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =sum(|0|)
const _1 = range(deps[0]);
return ctx['SUM'](_1);
}"
`;

exports[`expression compiler expression with $ref 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|0|+|1|+|2|
const _1 = ref(deps[0], false, "ADD",  undefined);
const _2 = ref(deps[1], false, "ADD",  undefined);
const _3 = ctx['ADD'](_1, _2);
const _4 = ref(deps[2], false, "ADD",  undefined);
return ctx['ADD'](_3, _4);
}"
`;

exports[`expression compiler expression with references with a sheet 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|0|
return ref(deps[0], false, "EQ",  undefined);
}"
`;

exports[`expression compiler expressions with a debugger 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =?|0|/|N0|
debugger;
const _1 = ref(deps[0], false, "DIVIDE",  undefined);
const _2 = { value: this.constantValues.numbers[0] };
return ctx['DIVIDE'](_1, _2);
}"
`;

exports[`expression compiler read some values and functions 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|0|+sum(|1|)
const _1 = ref(deps[0], false, "ADD",  undefined);
const _2 = range(deps[1]);
const _3 = ctx['SUM'](_2);
return ctx['ADD'](_1, _3);
}"
`;

exports[`expression compiler some arithmetic expressions 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|
return { value: this.constantValues.numbers[0] };
}"
`;

exports[`expression compiler some arithmetic expressions 2`] = `
"function anonymous(deps,ref,range,ctx
) {
// =true
return { value: true };
}"
`;

exports[`expression compiler some arithmetic expressions 3`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|S0|
return { value: this.constantValues.strings[0] };
}"
`;

exports[`expression compiler some arithmetic expressions 4`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|+|N1|
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
return ctx['ADD'](_1, _2);
}"
`;

exports[`expression compiler some arithmetic expressions 5`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|*|N1|
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
return ctx['MULTIPLY'](_1, _2);
}"
`;

exports[`expression compiler some arithmetic expressions 6`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|-|N1|
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
return ctx['MINUS'](_1, _2);
}"
`;

exports[`expression compiler some arithmetic expressions 7`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|/|N1|
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
return ctx['DIVIDE'](_1, _2);
}"
`;

exports[`expression compiler some arithmetic expressions 8`] = `
"function anonymous(deps,ref,range,ctx
) {
// =-|N0|
const _1 = { value: this.constantValues.numbers[0] };
return ctx['UMINUS'](_1);
}"
`;

exports[`expression compiler some arithmetic expressions 9`] = `
"function anonymous(deps,ref,range,ctx
) {
// =(|N0|+|N1|)*(-|N1|+|N2|)
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
const _3 = ctx['ADD'](_1, _2);
const _4 = { value: this.constantValues.numbers[1] };
const _5 = ctx['UMINUS'](_4);
const _6 = { value: this.constantValues.numbers[2] };
const _7 = ctx['ADD'](_5, _6);
return ctx['MULTIPLY'](_3, _7);
}"
`;

exports[`expression compiler some arithmetic expressions 10`] = `
"function anonymous(deps,ref,range,ctx
) {
// =sum(|N0|,|N1|)
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
return ctx['SUM'](_1,_2);
}"
`;

exports[`expression compiler some arithmetic expressions 11`] = `
"function anonymous(deps,ref,range,ctx
) {
// =sum(true,|S0|)
const _1 = { value: true };
const _2 = { value: this.constantValues.strings[0] };
return ctx['SUM'](_1,_2);
}"
`;

exports[`expression compiler some arithmetic expressions 12`] = `
"function anonymous(deps,ref,range,ctx
) {
// =sum(|N0|,,|N1|)
const _1 = { value: this.constantValues.numbers[0] };
const _2 = undefined;
const _3 = { value: this.constantValues.numbers[1] };
return ctx['SUM'](_1,_2,_3);
}"
`;

exports[`expression compiler some arithmetic expressions 13`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|N0|%
const _1 = { value: this.constantValues.numbers[0] };
return ctx['UNARY.PERCENT'](_1);
}"
`;

exports[`expression compiler some arithmetic expressions 14`] = `
"function anonymous(deps,ref,range,ctx
) {
// =(|N0|+|N1|)%
const _1 = { value: this.constantValues.numbers[0] };
const _2 = { value: this.constantValues.numbers[1] };
const _3 = ctx['ADD'](_1, _2);
return ctx['UNARY.PERCENT'](_3);
}"
`;

exports[`expression compiler some arithmetic expressions 15`] = `
"function anonymous(deps,ref,range,ctx
) {
// =|0|%
const _1 = ref(deps[0], false, "UNARY.PERCENT",  undefined);
return ctx['UNARY.PERCENT'](_1);
}"
`;

exports[`expression compiler with the same reference multiple times 1`] = `
"function anonymous(deps,ref,range,ctx
) {
// =SUM(|0|,|0|,|2|)
const _1 = range(deps[0]);
const _2 = range(deps[0]);
const _3 = range(deps[2]);
return ctx['SUM'](_1,_2,_3);
}"
`;
